### PR TITLE
feat: date_add and date_sub functions

### PR DESCRIPTION
**Pull request by mbutrovich**
## Which issue does this PR close?

Does not close an individual PR, but addresses an item in #858.

## What changes are included in this PR?

New DateAdd and DateSub functions in QueryPlanSerde. They have two arguments: a date and an int32 representing the number of days. DataFusion does not fully support `date_add`, so this PR introduces functions using the scalar UDF interface. These functions use DataFusion interfaces to apply Arrow kernels to perform the computation.

I also had to generate new q72 reference plans because it uses `date_add`.

## How are these changes tested?

New test cases for `date_add` and `date_sub`, both scalar and array arguments. `fix: ReusedExchangeExec + CometShuffleExchangeExec under QueryStageExec + should be CometRoot` in `CometExecSuite` also exercises `date_add`. q72 also exercises `date_add` as well.

### CODE REVIEW

**Comment by mbutrovich** [comment]
I'll drive into this test failure tomorrow. My assumption of the second expression for date_add always being a Literal does not appear sound.
```
Cause: org.apache.comet.CometNativeException: General execution error with reason: Not implemented: DateAdd { left: Some(Expr { expr_struct: Some(Literal(Literal { datatype: Some(DataType { type_id: Date, type_info: None }), is_null: false, value: Some(IntVal(0)) })) }), right: Some(Expr { expr_struct: Some(Cast(Cast { child: Some(Expr { expr_struct: Some(Remainder(Remainder { left: Some(Expr { expr_struct: Some(Bound(BoundReference { index: 0, datatype: Some(DataType { type_id: Int64, type_info: None }) })) }), right: Some(Expr { expr_struct: Some(If(IfExpr { if_expr: Some(Expr { expr_struct: Some(Eq(Equal { left: Some(Expr { expr_struct: Some(Literal(Literal { datatype: Some(DataType { type_id: Int64, type_info: None }), is_null: false, value: Some(LongVal(4)) })) }), right: Some(Expr { expr_struct: Some(Literal(Literal { datatype: Some(DataType { type_id: Int64, type_info: None }), is_null: false, value: Some(LongVal(0)) })) }) })) }), true_expr: Some(Expr { expr_struct: Some(Literal(Literal { datatype: Some(DataType { type_id: Int64, type_info: None }), is_null: true, value: None })) }), false_expr: Some(Expr { expr_struct: Some(Literal(Literal { datatype: Some(DataType { type_id: Int64, type_info: None }), is_null: false, value: Some(LongVal(4)) })) }) })) }), fail_on_error: false, return_type: Some(DataType { type_id: Int64, type_info: None }) })) }), datatype: Some(DataType { type_id: Int32, type_info: None }), timezone: "America/Los_Angeles", eval_mode: Legacy, allow_incompat: false })) }) }.
```

File: native/spark-expr/src/scalar_funcs.rs [hunk 1]
```diff
@@ -547,3 +551,32 @@ pub fn spark_isnan(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionEr
...
+/// Spark-compatible `date_add` expression, which assumes days for the second argument, but we
+/// cannot directly add that to a Date32. We generate an IntervalDayTime from the second argument,
+/// and use DataFusion's interface to apply Arrow's add operator.
+pub fn spark_date_add(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionError> {
+    let start = &args[0];
+    if let ColumnarValue::Scalar(ScalarValue::Int32(Some(days))) = &args[1] {
+        let interval = IntervalDayTime::new(*days, 0);
+        let interval_cv = ColumnarValue::Scalar(ScalarValue::IntervalDayTime(Some(interval)));
+        let result = datum::apply(start, &interval_cv, add)?;
+        return Ok(result);
```
Comment (andygrove) [inline native/spark-expr/src/scalar_funcs.rs:564]: No need to unwrap the result with `?` and then wrap again in `Ok`.

```suggestion
        return datum::apply(start, &interval_cv, add);
```

File: native/spark-expr/src/scalar_funcs.rs [hunk 2]
```diff
@@ -547,3 +551,32 @@ pub fn spark_isnan(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionEr
...
+        for day in days.as_primitive::<Int32Type>().into_iter() {
+            if let Some(non_null_day) = day {
+                interval_builder.append_value(IntervalDayTime::new(non_null_day, 0));
+            } else {
+                interval_builder.append_null();
+            }
+        }
+        let interval_cv = ColumnarValue::Array(Arc::new(interval_builder.finish()));
+        let result = datum::apply(start, &interval_cv, add)?;
+        return Ok(result);
```
Comment (andygrove) [inline native/spark-expr/src/scalar_funcs.rs:576]: Same comment re unwrapping result:

```suggestion
        return datum::apply(start, &interval_cv, add);
```

File: native/spark-expr/src/scalar_funcs.rs [hunk 3]
```diff
@@ -547,3 +551,32 @@ pub fn spark_isnan(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionEr
...
+/// and use DataFusion's interface to apply Arrow's add operator.
+pub fn spark_date_add(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionError> {
+    let start = &args[0];
+    if let ColumnarValue::Scalar(ScalarValue::Int32(Some(days))) = &args[1] {
+        let interval = IntervalDayTime::new(*days, 0);
+        let interval_cv = ColumnarValue::Scalar(ScalarValue::IntervalDayTime(Some(interval)));
+        let result = datum::apply(start, &interval_cv, add)?;
+        return Ok(result);
+    } else if let ColumnarValue::Array(days) = &args[1] {
+        let mut interval_builder = IntervalDayTimeBuilder::new();
```
Comment (andygrove) [inline native/spark-expr/src/scalar_funcs.rs:566]: Because we know the size upfront, we can use `with_capacity` here instead of `new`.

```suggestion
        let mut interval_builder = IntervalDayTimeBuilder::with_capacity(days.len());
```

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala [hunk 4]
```diff
@@ -146,6 +146,32 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
...
   }
 
+  test("date_add with scalar") {
+    Seq(true, false).foreach { dictionaryEnabled =>
+      withTempDir { dir =>
+        val path = new Path(dir.toURI.toString, "test.parquet")
+        makeParquetFileAllTypes(path, dictionaryEnabled = dictionaryEnabled, 10000)
+        withParquetTable(path.toString, "tbl") {
+          checkSparkAnswerAndOperator(
+            "SELECT _2 FROM tbl WHERE _20 + 2 > CAST('1970-01-02' AS DATE)")
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:156]: A simpler approach for testing here may be to just use the expression in a projection. For example:

```sql
SELECT _20 + 2 FROM tbl
```

I also tried testing with subtracting instead of adding and the test failed because we don't support `date_sub` (I had assumed that `date_add` would accept negative values). We could add `date_sub` as a part of this PR or as a separate PR.

File: spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala [hunk 5]
```diff
@@ -146,6 +146,32 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
...
+        }
+      }
+    }
+  }
+
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:161]: It would be good to add a test for overflow as well. Something like this:

```scala
  test("date_add with scalar overflow") {
    Seq(true, false).foreach { dictionaryEnabled =>
      withTempDir { dir =>
        val path = new Path(dir.toURI.toString, "test.parquet")
        makeParquetFileAllTypes(path, dictionaryEnabled = dictionaryEnabled, 10000)
        withParquetTable(path.toString, "tbl") {
          val (sparkErr, cometErr) = checkSparkMaybeThrows(
            sql(s"SELECT _20 + ${Int.MaxValue} FROM tbl"))
          assert(sparkErr.get.getMessage.contains("integer overflow"))
          assert(cometErr.get.getMessage.contains("`NaiveDate + TimeDelta` overflowed"))
        }
      }
    }
  }
```

**Comment by mbutrovich** [comment]
q72 comparison (only SF 1 locally), last two rows of each (using Comet Exec) are relevant. main branch:
```
TPCDS Snappy:                             Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
q72                                                7047           7848        1133          2.2         459.1       1.0X
q72: Comet (Scan)                                  6824           7291         660          2.2         444.6       1.0X
q72: Comet (Scan, Exec)                           10816          10896         112          1.4         704.7       0.7X
q72: Comet (Exec)                                 10811          10959         209          1.4         704.4       0.7X
```
This PR:
```
TPCDS Snappy:                             Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
q72                                                6848           6992         204          2.2         446.2       1.0X
q72: Comet (Scan)                                  6814           6819           7          2.3         444.0       1.0X
q72: Comet (Scan, Exec)                           10376          10442          93          1.5         676.1       0.7X
q72: Comet (Exec)                                 10472          10513          58          1.5         682.3       0.7X
```

File: native/spark-expr/src/scalar_funcs.rs [hunk 6]
```diff
@@ -547,3 +551,40 @@ pub fn spark_isnan(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionEr
...
+                interval_builder.append_value(IntervalDayTime::new(non_null_day, 0));
+            } else {
+                interval_builder.append_null();
+            }
+        }
+        let interval_cv = ColumnarValue::Array(Arc::new(interval_builder.finish()));
+        return datum::apply(start, &interval_cv, op);
+    }
+    Err(DataFusionError::Internal(format!(
+        "Unsupported data type {:?} for function date_add",
```
Comment (andygrove) [inline native/spark-expr/src/scalar_funcs.rs:580]: nit: hard-coded reference to `date_add` in error message but this is now also used for `date_sub`.

**Review by andygrove — APPROVED** [review]
LGTM with one nit. Thanks @mbutrovich!

File: native/spark-expr/src/scalar_funcs.rs [hunk 7]
```diff
@@ -547,3 +551,40 @@ pub fn spark_isnan(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionEr
...
+
+/// Spark-compatible `date_add` and `date_sub` expressions, which assumes days for the second
+/// argument, but we cannot directly add that to a Date32. We generate an IntervalDayTime from the
+/// second argument and use DataFusion's interface to apply Arrow's operators.
+fn spark_date_arithmetic(
+    args: &[ColumnarValue],
+    op: impl Fn(&dyn Datum, &dyn Datum) -> Result<ArrayRef, ArrowError>,
+) -> Result<ColumnarValue, DataFusionError> {
+    let start = &args[0];
+    if let ColumnarValue::Scalar(ScalarValue::Int32(Some(days))) = &args[1] {
```
Comment (viirya) [inline native/spark-expr/src/scalar_funcs.rs:563]: Hmm, I think Spark `DateAdd`'s second parameter could be IntegerType, ShortType, ByteType, i.e., Int32, Int16, Int8 in arrow types.

Comment (mbutrovich) [reply native/spark-expr/src/scalar_funcs.rs:563]: Good catch!

**Comment by mbutrovich** [comment]
I added support for Int8 and Int16. I don't love this control flow. @viirya suggested macros, but I'm open for other ways to approach this.

**Review by kazuyukitanimura — APPROVED** [review]

File: native/spark-expr/src/scalar_funcs.rs [hunk 8]
```diff
@@ -547,3 +551,76 @@ pub fn spark_isnan(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionEr
...
 }
+
+macro_rules! scalar_date_arithmetic {
+    ($start:expr, $days:expr, $op:expr) => {{
+        let interval = IntervalDayTime::new(*$days as i32, 0);
+        let interval_cv = ColumnarValue::Scalar(ScalarValue::IntervalDayTime(Some(interval)));
+        datum::apply($start, &interval_cv, $op)
+    }};
+}
+macro_rules! array_date_arithmetic {
```
Comment (viirya) [inline native/spark-expr/src/scalar_funcs.rs:562]: nit: style

```suggestion
}

macro_rules! array_date_arithmetic {
```

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Thanks @mbutrovich @andygrove @kazuyukitanimura
